test/path_tokenizer.h

```diff
@@ -0,0 +1,56 @@
+//
+// Created by wkl04 on 12-12-2019.
+//
+
+#ifndef MQTTSUBSCRIPTION_PATH_TOKENIZER_H
+#define MQTTSUBSCRIPTION_PATH_TOKENIZER_H
+
+#include <algorithm>
+
+template<typename Output>
+inline void mqtt_path_tokenizer(MQTT_NS::string_view str, Output write)
+{
+    auto first = str.begin();
+    auto last = str.end();
+
+    while (true) {
+        auto pos = std::find(first, last, '/');
+        if(!write(MQTT_NS::string_view(first, std::distance(first, pos)))) {
+            break;
+        }
+
+        if (pos == last) {
+            break;
+        }
+
+        first = std::next(pos);
+    }
+}
+
+// See if the topic is valid (does not contain # or +)
+inline bool mqtt_valid_topic(MQTT_NS::string_view topic)
+{
+    if(topic.empty()) {
+        return false;
+    }
+
+    if(topic[0] == '$') {
```

Comment:
`if(topic.empty() || (topic[0] == '$')`

Reply:
The problem i have with this, it is dependent on the order of the '||' validation if this will crash. If topic[0] == '$' is evaluated first, on an empty string, the result is undefined.

Reply:
What about:

inline bool mqtt_valid_topic(MQTT_NS::string_view topic)
{
    return topic.empty() ? false : 
        (topic[0] != '$' && (topic.find_first_of("#+") == std::string::npos));
}

Reply:
C++ guarantees that || and && are evaluated left to right for all implementations.

Or am I mistaken on that somehow?

Reply:
Yes, I guess the evaluation order is guaranteed left-to-right. The standard says there is a sequence point after ||